### PR TITLE
Add and improve tests for API

### LINKED ISSUE #131: Add tests covering the content of GeoJSON properties
Since #100, the result (GeoJSON properties) were lacking the report metadata and results. This was fixed in #130. To prevent an issue like that in the future tests should be added to check if all necessary (or at least the ones read by the website) properties are set.

Probably using `get_report_properties_template()` would solve a lot of the problems.

### PR DESCRIPTION
**Pull request by matthiasschaub**
### Description

Add test cases covering the content of GeoJSON properties.

Improve API response schemata by using less logic to create the
schemata.

### Corresponding issue
Closes #131 

### Checklist
- [x] I have updated my branch to `main` (e.g. through `git rebase main`)
- [x] My code follows the [style guide](https://github.com/GIScience/ohsome-quality-analyst/blob/main/CONTRIBUTING.md#style-guide) and was checked with [pre-commit](https://github.com/GIScience/ohsome-quality-analyst/blob/main/CONTRIBUTING.md#pre-commit) before committing
- [x] I have commented my code
- [x] I have added sufficient unit and integration [tests](https://github.com/GIScience/ohsome-quality-analyst/blob/main/docs/development_setup.md#tests)
- [x] I have updated the [CHANGELOG.md](https://github.com/GIScience/ohsome-quality-analyst/blob/main/CHANGELOG.md)

### CODE REVIEW

File: workers/tests/integrationtests/api_response_schema.py [hunk 1]
```diff
@@ -1,42 +1,24 @@
+"""This module defines schemata for API responses.
+
+One API response should adhere to at least two schemata.
```
Comment (Hagellach37) [inline workers/tests/integrationtests/api_response_schema.py:3]: Hey,
here we might write a bit easier to understand. Is the following correct:

"Every API response must adhere to the general schema. Additionally API responses need to adhere to one additional schema depending on which endpoint is used (indicator, report)."

File: workers/tests/integrationtests/api_response_schema.py [hunk 2]
```diff
@@ -49,19 +31,75 @@ def get_response_schema() -> Schema:
     )
 
 
-def get_feature_schema(iterator: Optional[str] = None):
+def get_result_schema() -> Schema:
+    """Response schema for all endpoints except of 'indicator' and 'report' endpoints"""
```
Comment (Hagellach37) [inline workers/tests/integrationtests/api_response_schema.py:35]: I'm a bit confused here. Which other endpoints are this?

**Review by Hagellach37 — COMMENTED** [review]
Hey,
just a minor comment

File: workers/tests/integrationtests/api_response_schema.py (hunk 2 above)
Comment (matthiasschaub) [reply workers/tests/integrationtests/api_response_schema.py:35]: PR #106 added API-endpoints to list the names of indicators, reports, layers, datasets and feature id fields. Those will be a list at the response key `result`. Those will be available in production once we release version 0.6.0.

Comment (matthiasschaub) [reply workers/tests/integrationtests/api_response_schema.py:35]: And there is also the `regions` endpoint to retrive all regions of dataset regions as GeoJSON.

File: workers/tests/integrationtests/api_response_schema.py (hunk 1 above)
Comment (matthiasschaub) [reply workers/tests/integrationtests/api_response_schema.py]: Yes thats correct. I used your formulation and updated the PR.

**Comment by matthiasschaub** [comment]
@Hagellach37 please feel free to merge this branch if it nothing more needs to be changed.

File: workers/tests/integrationtests/api_response_schema.py [hunk 3]
```diff
@@ -49,19 +32,75 @@ def get_response_schema() -> Schema:
...
+        "result.description": str,
+        "result.svg": str,
+    }
+    properties = {}
+    for i in range(number_of_indicators):
+        for k, v in properties_template.items():
+            if k == "result.svg":
+                properties.update({Opt("indicators." + str(i) + "." + k): v})
+            else:
+                properties.update({"indicators." + str(i) + "." + k: v})
```
Comment (joker234) [inline workers/tests/integrationtests/api_response_schema.py:89]: Why do you double the `properties_template` code here, if you overwrite `"result.svg"` anyway?

File: workers/tests/integrationtests/test_api_report_geojson_io.py [hunk 4]
```diff
@@ -39,30 +36,21 @@ def setUp(self):
...
-                self.validate(feature, feature_schema)
-
-    def validate(self, geojson: dict, schema: Schema) -> None:
-        schema.validate(geojson)  # Print information if validation fails
-        self.assertTrue(schema.is_valid(geojson))
+
+        response_content = geojson.loads(response.content)
+        self.assertTrue(response_content.is_valid)  # Valid GeoJSON?
+        self.assertTrue(self.general_schema.is_valid(response_content))
+        self.assertTrue(self.feature_schema.is_valid(response_content))
```
Comment (joker234) [inline workers/tests/integrationtests/test_api_report_geojson_io.py:53]: is the `feature_schema` handling having more indicators than defined?

**Review by joker234 — COMMENTED** [review]
:+1: Only two minor questions.

File: workers/tests/integrationtests/api_response_schema.py (hunk 3 above)
Comment (joker234) [reply workers/tests/integrationtests/api_response_schema.py:89]: This code smell is related to this issue: https://sonarcloud.io/project/issues?branch=PR-168&id=ohsome-quality-analyst&open=AXu650vLVi-dpvTMVG1C&resolved=false&types=CODE_SMELL

File: workers/tests/integrationtests/test_api_report_geojson_io.py (hunk 4 above)
Comment (matthiasschaub) [reply workers/tests/integrationtests/test_api_report_geojson_io.py:53]: In this case `self.feature_schema` is derived from calling `get_report_feature_schema(number_of_indicators)`. If the response contains more indicators then the `numbers_of_indicators` passed to `get_report_feature_schema` the validity check will fail.

Does this answer your question?

File: workers/tests/integrationtests/api_response_schema.py [hunk 5]
```diff
@@ -49,19 +41,54 @@ def get_response_schema() -> Schema:
...
-        {"type": "FeatureCollection", "features": list}, ignore_extra_keys=True
+def get_report_feature_schema(number_of_indicators: int) -> Schema:
+    properties_template = get_indicator_properties_template()
+    properties = {}
+    for i in range(number_of_indicators):
+        for k, v in properties_template.items():
+            if k == Opt("result.svg"):
+                properties.update({Opt("indicators." + str(i) + "." + "result.svg"): v})
+            else:
+                properties.update({"indicators." + str(i) + "." + k: v})
```
Comment (joker234) [inline workers/tests/integrationtests/api_response_schema.py:77]: I'm not a fan of these lines. Do you know a better way to handle this? Maybe something like this:
```suggestion
            prefix = "indicators." + str(i) + "."
            if isinstance(k, Schema):
                k._prepend_schema_name(prefix)
            else:
                k = prefix + k
            properties.update({"indicators." + str(i) + "." + k: v})
```
But this is just a *suggestion*, not something that needs to be changed.

File: workers/tests/integrationtests/test_api_report_geojson_io.py (hunk 4 above)
Comment (joker234) [reply workers/tests/integrationtests/test_api_report_geojson_io.py:53]: yes, thanks

**Review by joker234 — APPROVED** [review]

**Pull request merged**